kubernetes/cache.go

```diff
@@ -0,0 +1,325 @@
+	}
+	if len(pods) > 0 {
+		_, ok := pods[0].(*v1.Pod)
+		if !ok {
+			return nil, false
+		}
+		nsPods := make([]v1.Pod, len(pods))
+		for i, pod := range pods {
+			nsPods[i] = *(pod.(*v1.Pod))
+		}
+		return nsPods, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetReplicationControllers(namespace string) ([]v1.ReplicationController, bool) {
+	indexer := c.controllers["ReplicationController"].GetIndexer()
+	repcons, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
+		return nil, false
+	}
+	if len(repcons) > 0 {
+		_, ok := repcons[0].(*v1.ReplicationController)
+		if !ok {
+			return nil, false
+		}
+		nsRepcons := make([]v1.ReplicationController, len(repcons))
+		for i, repcon := range repcons {
+			nsRepcons[i] = *(repcon.(*v1.ReplicationController))
+		}
+		return nsRepcons, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetReplicaSets(namespace string) ([]v1beta2.ReplicaSet, bool) {
+	indexer := c.controllers["ReplicaSet"].GetIndexer()
+	repsets, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
+		return nil, false
+	}
+	if len(repsets) > 0 {
+		_, ok := repsets[0].(*v1beta2.ReplicaSet)
+		if !ok {
+			return nil, false
+		}
+		nsRepsets := make([]v1beta2.ReplicaSet, len(repsets))
+		for i, repset := range repsets {
+			nsRepsets[i] = *(repset.(*v1beta2.ReplicaSet))
+		}
+		return nsRepsets, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetStatefulSet(namespace, name string) (*v1beta2.StatefulSet, bool) {
+	indexer := c.controllers["StatefulSet"].GetIndexer()
+	fulsets, exist, err := indexer.GetByKey(namespace + "/" + name)
+	if err != nil {
+		return nil, false
+	}
+	if exist {
+		fulset, ok := fulsets.(*v1beta2.StatefulSet)
+		if !ok {
+			return nil, false
+		}
+		return fulset, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetStatefulSets(namespace string) ([]v1beta2.StatefulSet, bool) {
+	indexer := c.controllers["StatefulSet"].GetIndexer()
+	fulsets, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
+		return nil, false
+	}
+	if len(fulsets) > 0 {
+		_, ok := fulsets[0].(*v1beta2.StatefulSet)
+		if !ok {
+			return nil, false
+		}
+		nsFulsets := make([]v1beta2.StatefulSet, len(fulsets))
+		for i, fulset := range fulsets {
+			nsFulsets[i] = *(fulset.(*v1beta2.StatefulSet))
+		}
+		return nsFulsets, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetService(namespace, name string) (*v1.Service, bool) {
+	indexer := c.controllers["Service"].GetIndexer()
+	services, exist, err := indexer.GetByKey(namespace + "/" + name)
+	if err != nil {
+		return nil, false
+	}
+	if exist {
+		service, ok := services.(*v1.Service)
+		if !ok {
+			return nil, false
+		}
+		return service, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetServices(namespace string) ([]v1.Service, bool) {
+	indexer := c.controllers["Service"].GetIndexer()
+	services, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
```

Comment:
what kind of error could we have here?
maybe it's worth logging, and even sending as a result?

Reply:
I'm a going to check.
I did some tests to validate the recovery cappabilities (error propagation when backend is off, reconnect, etc) but I don't remember if this is missing.
I will update my comment after a couple of tests.